### PR TITLE
Update governance endpoint

### PR DESCRIPTION
**Pull request by danielr18**

### CODE REVIEW

**Comment by ariel-dao** [comment]
This looks like the correct way to approach it, you'll need to iterate over the common members (subcollection in common) and update their circle ids accordingly if needed

**Comment by danielr18** [comment]
@ariel-dao Can you review it again?
I modified the endpoint to preserve ids

File: functions/src/governance/controllers/update.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { USER_ROLES } from '../../shared/enums/USER_ROLES';
+import { CommonError } from '../../util/errors';
+
+export const update = async (req: Request, res: Response, next: NextFunction) => {
+    await responseExecutor(
+        async () => {
+            const { user: { uid: userId } } = req;
+            const { roles } = await userDb.get(userId, true);
+            const isAdmin = (roles || []).some(role => role === USER_ROLES.ADMIN);
+            if (!isAdmin) throw new CommonError('User without `admin` role tried to update governance', req);
```
Comment (ariel-dao) [inline functions/src/governance/controllers/update.ts]: You can just throw Unauthorized, no need to share info like that (you can log it using logger.error)

File: functions/src/governance/services/update.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { governanceDb } from '../db';
+import { Governance } from '../../shared/interfaces/governance/Governance';
+
+export const update = async (change: Pick<Governance, 'actions' | 'circles' | 'proposals'>, governanceId: string) => {
+    // @ts-expect-error - there is an issue with the recursive partial type
```
Comment (ariel-dao) [inline functions/src/governance/services/update.ts]: We need to fix this error, can you debug it?

File: functions/src/governance/db/update.ts [hunk 3]
```diff
@@ -6,16 +6,24 @@ import { get } from './get';
...
-export const updateGovernance = async (governanceId: string, governanceUpdate: UpdatableGovernanceFields) => {
+export const updateGovernance = async (governanceId: string, governanceUpdate: UpdatableGovernanceFields, settings?: UpdateGovernanceSettings) => {
     await get({ governanceId }, true);
 
     const governanceDoc = { ...governanceUpdate, updatedAt: Timestamp.now(), };
 
-    await GovernanceCollection.doc(governanceId).set(governanceDoc as Governance, { merge: true });
+    const updateMethod = settings?.method ?? 'merge';
+    if (updateMethod === 'update') {
+        // @ts-expect-error - there is an issue with the recursive partial type
```
Comment (ariel-dao) [inline functions/src/governance/db/update.ts]: Need to fix this error

File: functions/src/governance/db/update.ts [hunk 4]
```diff
@@ -6,16 +6,24 @@ import { get } from './get';
...
+        await GovernanceCollection.doc(governanceId).update(governanceDoc);
+    } else if (updateMethod === 'merge') {
+        await GovernanceCollection.doc(governanceId).set(governanceDoc as Governance, { merge: true });
```
Comment (ariel-dao) [inline functions/src/governance/db/update.ts]: I'm not sure you want to use merge true for this type of update - what if something is deleted?

File: functions/src/governance/business/governanceValidationAndFormat.ts [hunk 5]
```diff
@@ -0,0 +1,327 @@
+import { v4 } from 'uuid';
```
Comment (ariel-dao) [inline functions/src/governance/business/governanceValidationAndFormat.ts]: I understand the renaming, but it really makes it hard to spot the differences if there are any - just fyi (maybe changing the name back just for this PR would make it better)

**Review by ariel-dao — CHANGES_REQUESTED** [review]
Looks like the right direction, we need to update the circle Ids array accordingly in case circles are being changed (check `CommonMember`). 

Also we need to recalculate proposal votes and change the weights on **live** proposals, See `onRecalculateTotalVotingPower` and `assignCircleToMemberAction`, the process is already defined so you don't need to deal with the complexity. You'd probably need to emit an event that would get caught by `onRecalculateTotalVotingPower`.

Another thing, please keep the branch updated with the base branch.

File: functions/src/governance/db/update.ts (hunk 4 above)
Comment (danielr18) [reply functions/src/governance/db/update.ts]: Merge true was the method used before this PR, that's why I kept it as the default method, and use the `update` method in the new controller.

Do you think we should only keep update?

Comment (ariel-dao) [reply functions/src/governance/db/update.ts]: I think we should have another update function that won't merge, just duplicate the update and remove the merge for simple solution

File: functions/src/governance/services/update.ts (hunk 2 above)
Comment (danielr18) [reply functions/src/governance/services/update.ts]: Fixed

File: functions/src/governance/db/update.ts (hunk 4 above)
Comment (danielr18) [reply functions/src/governance/db/update.ts]: I've created another update function

File: functions/src/governance/db/update.ts (hunk 3 above)
Comment (danielr18) [reply functions/src/governance/db/update.ts]: Fixed

**Comment by danielr18** [comment]
> Looks like the right direction, we need to update the circle Ids array accordingly in case circles are being changed (check `CommonMember`).

The Ids stay the same, so the CommonMember circleIds may not need to be updated.

> Also we need to recalculate proposal votes and change the weights on **live** proposals, See `onRecalculateTotalVotingPower` and `assignCircleToMemberAction`, the process is already defined so you don't need to deal with the complexity. You'd probably need to emit an event that would get caught by `onRecalculateTotalVotingPower`.
> 
> Another thing, please keep the branch updated with the base branch.

I've created a new event to trigger this function.

**Review by ariel-dao — APPROVED** [review]

**Pull request merged**